AdventOfCode23/Day3/Schematic.cs

```diff
@@ -0,0 +1,79 @@
+﻿using System.Text.RegularExpressions;
+
+namespace AdventOfCode23.Day3;
+
+public static partial class Schematic
+{
+    public static int SumAllPartNumbers(IEnumerable<string> data)
+    {
+        int sum = 0;
+        var dataArray = data.ToArray();
+        for (var i = 0; i < dataArray.Length; i++)
+        {
+            for (int j = 0; j < dataArray[i].Length; j++)
+            {
+                int start = j;
+                while (j < dataArray[i].Length && int.TryParse(dataArray[i][j].ToString(), out _))
+                {
+                    j++;
+                }
+
+                if (start == j)
+                {
+                    continue;
+                }
+
+                j--;
+
+                int searchStart = start > 0 ? start - 1 : start;
+                int searchEnd = j < dataArray[i].Length - 1 ? j + 1 : j;
+                string surrChars;
+                switch (i)
+                {
+                    case > 0 when i + 1 < dataArray.Length:
+                        surrChars = string.Concat(
+                            dataArray[i - 1].Substring(searchStart, searchEnd - searchStart + 1),
+                            searchStart == start ? "" : dataArray[i][start - 1].ToString(),
+                            searchEnd == j ? "" : dataArray[i][j + 1].ToString(),
+                            dataArray[i + 1].Substring(searchStart, searchEnd - searchStart + 1)
+                        );
+                        if (MyRegex().IsMatch(surrChars))
+                        {
+                            sum += int.Parse(dataArray[i].Substring(start, j - start + 1));
+                        }
+
+                        break;
+                    case 0:
+                        surrChars = string.Concat(
+                            searchStart == start ? "" : dataArray[i][start - 1].ToString(),
+                            searchEnd == j ? "" : dataArray[i][j + 1].ToString(),
+                            dataArray[i + 1].Substring(searchStart, searchEnd - searchStart + 1)
+                        );
+                        if (MyRegex().IsMatch(surrChars))
+                        {
+                            sum += int.Parse(dataArray[i].Substring(start, j - start + 1));
+                        }
+
+                        break;
+                    default:
+                        surrChars = string.Concat(
+                            dataArray[i - 1].Substring(searchStart, searchEnd - searchStart + 1),
+                            searchStart == start ? "" : dataArray[i][start - 1].ToString(),
+                            searchEnd == j ? "" : dataArray[i][j + 1].ToString()
+                        );
+                        if (MyRegex().IsMatch(surrChars))
+                        {
+                            sum += int.Parse(dataArray[i].Substring(start, j - start + 1));
+                        }
+
+                        break;
```

Comment:
Ok, look... I know this isn't the prettiest thing you have ever seen... I should've done something fancier with maybe some sort of function which generates a map of the surrounding indexes, traverses it, and returns back the surrounding char string... But it just so happens that I know on Part 2, none of that would be useful, so I didn't really want to spend too much time on it.

Once I'm done with AoC, I plan to come back and revisit each of the problems, seeing if I can improve the functions at all. We'll see what we can do at that point.